src/cache/inmemory/readFromStore.ts

```diff
@@ -272,30 +273,28 @@ export class StoreReader {
       },
     });
 
-    let missing: MissingFieldError[] | undefined;
+    let missing: MissingFieldError | undefined;
     if (execResult.missing) {
-      // For backwards compatibility we still report an array of
-      // MissingFieldError objects, even though there will only ever be at most
-      // one of them, now that all missing field error messages are grouped
-      // together in the execResult.missing tree.
-      missing = [
-        new MissingFieldError(
-          firstMissing(execResult.missing)!,
-          execResult.missing,
-          query,
-          variables
-        ),
-      ];
-      if (!returnPartialData) {
-        throw missing[0];
-      }
+      missing = new MissingFieldError(
+        firstMissing(execResult.missing)!,
+        execResult.missing,
+        query,
+        variables
+      );
     }
 
+    const complete = !missing;
+    const { result } = execResult;
+
     return {
-      result: execResult.result,
-      complete: !missing,
+      result:
+        complete || returnPartialData ?
+          equal(result, {}) ? null
```

Comment:
```suggestion
          Object.keys(result).length === 0 ?
            null
```
`equal` does a lot of lifting, do you think we could get away with something like this too?

Reply:
Works for me! [e70c3fc](https://github.com/apollographql/apollo-client/pull/12304/commits/e70c3fc887f3c2090ca5b51dfa1ef94c806e1ff9)